tiledb/common/memory_tracker.cc

```diff
@@ -0,0 +1,319 @@
+   public:
+    MemoryTrackerCreator()
+        : MemoryTracker() {
+    }
+  };
+
+  std::lock_guard<std::mutex> lg(mutex_);
+
+  // Delete any expired weak_ptr instances
+  size_t idx = 0;
+  while (idx < trackers_.size()) {
+    if (trackers_[idx].expired()) {
+      trackers_.erase(trackers_.begin() + idx);
+    } else {
+      idx++;
+    }
+  }
+
+  // Create a new tracker
+  auto ret = make_shared<MemoryTrackerCreator>(HERE());
+  trackers_.emplace(trackers_.begin(), ret);
+
+  return ret;
+}
+
+std::string MemoryTrackerManager::to_json() {
+  std::lock_guard<std::mutex> lg(mutex_);
+  nlohmann::json rv;
+
+  size_t idx = 0;
+  while (idx < trackers_.size()) {
+    auto ptr = trackers_[idx].lock();
+    if (!ptr) {
+      trackers_.erase(trackers_.begin() + idx);
+      continue;
+    }
+
+    nlohmann::json val;
+
+    // Set an distinguishing id
+    val["tracker_id"] = std::to_string(ptr->get_id());
+
+    // Mark the stats with the tracker type.
+    val["tracker_type"] = memory_tracker_type_to_str(ptr->get_type());
+
+    // Add memory stats
+    auto [total, by_type] = ptr->get_counts();
+    val["total_memory"] = total;
+    val["by_type"] = nlohmann::json::object();
+    for (auto& [type, count] : by_type) {
+      val["by_type"][memory_type_to_str(type)] = count;
+    }
+    rv.push_back(val);
+
+    idx++;
+  }
+
+  return rv.dump();
+}
+
+MemoryTrackerReporter::~MemoryTrackerReporter() {
+  if (!filename_.has_value()) {
+    return;
+  }
+
+  {
+    std::lock_guard<std::mutex> lg(mutex_);
+    stop_ = true;
+    cv_.notify_all();
+  }
+  try {
+    thread_.join();
+  } catch (std::exception& exc) {
+    LOG_ERROR(
+        "Error stopping MemoryTrackerReporter thread: " +
+        std::string(exc.what()));
+  }
+}
+
+void MemoryTrackerReporter::start() {
+  if (!filename_.has_value()) {
+    LOG_INFO("No filename set, not starting the MemoryTrackerReporter.");
+    return;
+  }
+
+  {
+    // Scoped so we release this before the thread starts. Probably unnecessary
+    // by better safe than sorry.
+    std::lock_guard<std::mutex> lg(mutex_);
+    if (stop_) {
+      throw std::runtime_error("MemoryTrackerReporters cannot be restarted.");
+    }
+  }
+
+  // Thread start logic mirrored from the ThreadPool.
+  for (size_t i = 0; i < 3; i++) {
+    try {
+      thread_ = std::thread(&MemoryTrackerReporter::run, this);
+      return;
+    } catch (const std::system_error& e) {
+      if (e.code() == std::errc::resource_unavailable_try_again) {
+        continue;
+      }
+
+      LOG_ERROR(
+          "Error starting the MemoryTrackerReporter: " + std::string(e.what()));
+    }
+  }
+
+  LOG_ERROR("Failed to start the MemoryTrackerReporter thread.");
+}
+
+void MemoryTrackerReporter::run() {
+  std::stringstream ss;
+  std::ofstream out;
+  out.open(filename_.value(), std::ios::app);
+
+  if (!out) {
+    LOG_ERROR("Error opening MemoryTrackerReporter file: " + filename_.value());
```

Comment:
Same comment here as above for throwing.

Reply:
This one I'm slightly less inclined to turn into an exception. The first reason is that throwing an exception here just translates into a `std::terminate` since it's not handled by the thread. Secondly, blowing up because of things changing out from under us at runtime could be a weird behavior.

However, I'll freely admit to writing this in haste with just a goal of getting something working. I think a better approach would be to handle any errors that arise more gracefully and then handle resuming log writes when/if the file becomes writable again. Which is to say, keeping these as log messages but trying to resume writing when we're able is likely better than crashing the entire process.

Reply:
Let’s file a follow up story to clean this up if needed…

Reply:
I kept the log messages but added the logic to continually attempt re-opening the log instead of just silently closing the reporter on first error.